File: android/src/main/kotlin/com/changteng/agora_rtc/AgoraRtcEngineRecordingActions.kt
package com.changteng.agora_rtc

import io.agora.rtc2.AgoraMediaRecorder
import io.agora.rtc2.IMediaRecorderCallback
import io.agora.rtc2.RecorderInfo
import io.agora.rtc2.RecorderStreamInfo
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

class AgoraRtcEngineRecordingActions(
  private val engineHolder: AgoraRtcEngineHolder,
  private val emitter: AgoraEventEmitter
) {
  private var mediaRecorder: AgoraMediaRecorder? = null
  private var shouldDestroyOnStop: Boolean = false

  fun clear() {
    mediaRecorder = null
    shouldDestroyOnStop = false
  }

  fun handleStartRecording(call: MethodCall, result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val args = call.arguments as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "参数缺失", null)
      return
    }
    val configMap = args["config"] as? Map<*, *> ?: run {
      result.error("INVALID_ARGS", "config 为空", null)
      return
    }
    val channelId = configMap["channelId"] as? String ?: ""
    val uid = (configMap["uid"] as? Number)?.toInt() ?: 0
    val storagePath = configMap["storagePath"] as? String
    val containerFormat = (configMap["containerFormat"] as? Number)?.toInt() ?: AgoraMediaRecorder.CONTAINER_MP4
    val streamType = (configMap["streamType"] as? Number)?.toInt() ?: AgoraMediaRecorder.STREAM_TYPE_BOTH
    val maxDurationMs = (configMap["maxDurationMs"] as? Number)?.toInt() ?: 120000
    val recorderInfoUpdateInterval = (configMap["recorderInfoUpdateInterval"] as? Number)?.toInt() ?: 0
    if (storagePath.isNullOrBlank()) {
      result.error("INVALID_ARGS", "storagePath 为空", null)
      return
    }
    try {
      if (mediaRecorder == null) {
        val streamInfo = RecorderStreamInfo().apply {
          this.channelId = channelId
          this.uid = uid
        }
        mediaRecorder = engine.createMediaRecorder(streamInfo)
        mediaRecorder?.setMediaRecorderObserver(object : IMediaRecorderCallback {
          override fun onRecorderStateChanged(channelId: String?, uid: Int, state: Int, error: Int) {
            emitter.emit(
              "onRecorderStateChanged",
              mapOf(
                "channelId" to (channelId ?: ""),
                "uid" to uid,
                "state" to state,
                "reason" to error
              )
            )
            if (shouldDestroyOnStop) {
              mediaRecorder?.setMediaRecorderObserver(null)
              engine.destroyMediaRecorder(mediaRecorder)
              mediaRecorder = null
              shouldDestroyOnStop = false
            }
          }

          override fun onRecorderInfoUpdated(channelId: String?, uid: Int, info: RecorderInfo?) {
          }
        })
      }
      val config = AgoraMediaRecorder.MediaRecorderConfiguration(
        storagePath,
        containerFormat,
        streamType,
        maxDurationMs,
        recorderInfoUpdateInterval
      )
      val code = mediaRecorder?.startRecording(config) ?: -1
      result.success(code)
    } catch (ex: Exception) {
      result.error("RECORDER_UNSUPPORTED", ex.message, null)
    }
  }

  fun handleStopRecording(result: Result) {
    val engine = engineHolder.rtcEngine
    if (engine == null) {
      result.error("NO_ENGINE", "RtcEngine 未初始化", null)
      return
    }
    val recorder = mediaRecorder ?: run {
      result.success(-1)
      return
    }
    shouldDestroyOnStop = true
    val stopCode = recorder.stopRecording()
    if (stopCode < 0) {
      recorder.setMediaRecorderObserver(null)
      engine.destroyMediaRecorder(recorder)
      mediaRecorder = null
      shouldDestroyOnStop = false
    }
    result.success(stopCode)
  }
}

File: ios/Classes/AgoraRtcEngineRecordingActions.swift
import Flutter
import AgoraRtcKit

final class AgoraRtcEngineRecordingActions: NSObject {
  private let engineHolder: AgoraRtcEngineViewHolder
  private let emitter: AgoraEventEmitter
  private var mediaRecorder: AgoraMediaRecorder?
  private var shouldDestroyOnStop: Bool = false

  init(engineHolder: AgoraRtcEngineViewHolder, emitter: AgoraEventEmitter) {
    self.engineHolder = engineHolder
    self.emitter = emitter
    super.init()
  }

  func clear() {
    mediaRecorder = nil
    shouldDestroyOnStop = false
  }

  func handleStartRecording(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "参数缺失", details: nil))
      return
    }
    guard let configMap = args["config"] as? [String: Any] else {
      result(FlutterError(code: "INVALID_ARGS", message: "config 为空", details: nil))
      return
    }
    let channelId = (configMap["channelId"] as? String) ?? ""
    let uid = (configMap["uid"] as? NSNumber)?.uintValue ?? 0
    guard let storagePath = configMap["storagePath"] as? String, !storagePath.isEmpty else {
      result(FlutterError(code: "INVALID_ARGS", message: "storagePath 为空", details: nil))
      return
    }
    let containerFormat = (configMap["containerFormat"] as? NSNumber)?.intValue
    let maxDurationMs = (configMap["maxDurationMs"] as? NSNumber)?.intValue ?? 120000
    if mediaRecorder == nil {
      let streamInfo = AgoraRecorderStreamInfo()
      streamInfo.channelId = channelId
      streamInfo.uid = uid
      mediaRecorder = engine.createMediaRecorder(withInfo: streamInfo)
      mediaRecorder?.setMediaRecorderDelegate(self)
    }
    guard let recorder = mediaRecorder else {
      result(FlutterError(code: "RECORDER_UNSUPPORTED", message: "创建录制实例失败", details: nil))
      return
    }
    let config = AgoraMediaRecorderConfiguration()
    config.storagePath = storagePath
    if let rawValue = containerFormat,
       let format = AgoraMediaRecorderContainerFormat(rawValue: rawValue) {
      config.containerFormat = format
    } else {
      config.containerFormat = .MP4
    }
    config.maxDurationMs = UInt(maxDurationMs)
    let code = recorder.startRecording(config)
    result(code)
  }

  func handleStopRecording(result: @escaping FlutterResult) {
    guard let engine = engineHolder.rtcEngine else {
      result(FlutterError(code: "NO_ENGINE", message: "RtcEngine 未初始化", details: nil))
      return
    }
    guard let recorder = mediaRecorder else {
      result(-1)
      return
    }
    shouldDestroyOnStop = true
    let stopCode = recorder.stopRecording()
    if stopCode < 0 {
      recorder.setMediaRecorderDelegate(nil)
      engine.destroy(recorder)
      mediaRecorder = nil
      shouldDestroyOnStop = false
    }
    result(stopCode)
  }
}

extension AgoraRtcEngineRecordingActions: AgoraMediaRecorderDelegate {
  func mediaRecorder(
    _ recorder: AgoraMediaRecorder,
    stateDidChanged channelId: String,
    uid: UInt,
    state: AgoraMediaRecorderState,
    reason: AgoraMediaRecorderReasonCode
  ) {
    emitter.emit(
      type: "onRecorderStateChanged",
      data: [
        "channelId": channelId,
        "uid": Int(uid),
        "state": state.rawValue,
        "reason": reason.rawValue,
      ]
    )
    if shouldDestroyOnStop {
      recorder.setMediaRecorderDelegate(nil)
      engineHolder.rtcEngine?.destroy(recorder)
      mediaRecorder = nil
      shouldDestroyOnStop = false
    }
  }

  func mediaRecorder(
    _ recorder: AgoraMediaRecorder,
    informationDidUpdated channelId: String,
    uid: UInt,
    info: AgoraMediaRecorderInfo
  ) {
  }
}
